storage: de-flake TestStoreRangeRemoveDead

tbg opened the pull request:
The test could end up with eternally throttled stores thanks to the use
of a manual clock. Configure the timeouts to something that is tiny
compared to the manual clock ticks performed in the test.

Fixes #32408.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/32806)
<!-- Reviewable:end -->

a-robinson reviewed (changes requested):
> Reviewed 5 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32806)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/client_test.go, line 299 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32806#-LSq38Pe1lRgQgYcyBJr:-LSq38Pe1lRgQgYcyBJs:bxtkdhn) ([raw file](https://github.com/cockroachdb/cockroach/blob/f811ae0d984d996792e9865faf18c9c11098346a/pkg/storage/client_test.go#L299)):*
> ```Go
> 	)
> 
> 	const basicallyForever = 24 * time.Hour
> ```

Why not use `TestTimeUntilStoreDeadOff`?
___
*[pkg/storage/client_test.go, line 304 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32806#-LSq32jqCGtk2wDFVajt:-LSq32jqCGtk2wDFVaju:bf2zq5l) ([raw file](https://github.com/cockroachdb/cockroach/blob/f811ae0d984d996792e9865faf18c9c11098346a/pkg/storage/client_test.go#L304)):*
> ```Go
> 	}
> 	if m.declinedReservationsTimeout == 0 {
> 		m.declinedReservationsTimeout = basicallyForever
> ```

Why would we want to decline/fail reservations forever by default? This seems like it would make a lot of tests flaky that previously relied on declined snapshots succeeding after a second.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
I was confused for a while why this test didn't repro the problem when I commented out my actual change to lower the timings for the declined and failed reservations. What I found out was that this test wasn't actually doing anything: in never gossiping the "dead" store, that store would never actually receive replicas, and so the "replicates away from dead node" check was trivially true. I fixed that in the test, good thing it had this unrelated failure mode or we probably would've never known.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32806#-:-LSt31pU2EQxkKo4uFLy:b3i0zxj)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/client_test.go, line 299 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32806#-LSq38Pe1lRgQgYcyBJr:-LSs9TAY7PtIZJdj6sud:b-i1rfjz) ([raw file](https://github.com/cockroachdb/cockroach/blob/f811ae0d984d996792e9865faf18c9c11098346a/pkg/storage/client_test.go#L299)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Why not use `TestTimeUntilStoreDeadOff`?
</blockquote></details>

I was hoping to get rid of it and did, but backed it out again to keep the diff small.
___
*[pkg/storage/client_test.go, line 304 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/32806#-LSq32jqCGtk2wDFVajt:-LSsABL08ieQHbKbm92v:bxu2g2z) ([raw file](https://github.com/cockroachdb/cockroach/blob/f811ae0d984d996792e9865faf18c9c11098346a/pkg/storage/client_test.go#L304)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Why would we want to decline/fail reservations forever by default? This seems like it would make a lot of tests flaky that previously relied on declined snapshots succeeding after a second.
</blockquote></details>

The multiTestContext typically uses a manual clock, which means that a timeout-based mechanism is going to backfire unless it's fully disabled by default, and I just prefer to leave the code paths active and to weed out the tests that do something shady. Neither option is particularly satisfying and really the clock would need to provide the whole API for measuring time durations (including some hatch for manual clocks), but that's not a yak I want to shave.
I also think this throttling is hit for too many reasons right now (this probably played a role in #31409, though now I've hit it with a large enough hammer), so I want it to be obvious about when it's happening.

I've backed this out though to keep the diff small.


<!-- Sent from Reviewable.io -->

tbg committed:
storage: de-flake TestStoreRangeRemoveDead

The test could end up with eternally throttled stores thanks to the use
of a manual clock. Configure the timeouts to something that is tiny
compared to the manual clock ticks performed in the test.

The test hits snapshot rejections because it (sometimes, automatically)
first upreplicates to, say, replicas 1 3 5 4 6 and then rapidly tries to
remove and re-add 5, before the old replica has been GC'ed, which runs
into #32436.

Fixes #32408.

Release note: None

a-robinson reviewed (approved):
> Reviewed 3 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32806)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/client_raft_test.go, line 2903 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32806#-LSuHu1eBSzRWxTdeLb9:-LSuHu1eBSzRWxTdeLbA:b4wcnqp) ([raw file](https://github.com/cockroachdb/cockroach/blob/a85fd382eb864ddd14031ff6c8daa683e79d0d78/pkg/storage/client_raft_test.go#L2903)):*
> ```
> 					// Until the deadStore dies, don't gossip one of the other
> 					// stores to force some replicas on deadStore.
> ```

I'm kind of surprised this is needed, I'd have thought that just adding the `deadStore.GossipStore` call would have been enough to ensure replication.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
bors r=a-robinson

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32806#-:-LSuMwh60ANfnP2h1oel:bd334lo)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/client_raft_test.go, line 2903 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/32806#-LSuHu1eBSzRWxTdeLb9:-LSuMlty3I9r0Hj_a6QS:b4nxy62) ([raw file](https://github.com/cockroachdb/cockroach/blob/a85fd382eb864ddd14031ff6c8daa683e79d0d78/pkg/storage/client_raft_test.go#L2903)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

> ```
> 					// Until the deadStore dies, don't gossip one of the other
> 					// stores to force some replicas on deadStore.
> ```

I'm kind of surprised this is needed, I'd have thought that just adding the `deadStore.GossipStore` call would have been enough to ensure replication.
</blockquote></details>

It's probably not needed (because the rebalancer will make sure the range counts are approximately even?). I'm going to merge this, but let me know if you want me to remove it. I just didn't want to take any chances.


<!-- Sent from Reviewable.io -->
